### ISSUE TITLE
[Bug] ValueError: Weight proof validation failed

### ISSUE THREAD

**Issue by teknohog** [Chia-Network/chia-blockchain#17632]
### What happened?

My full node lost sync in December 2023 and I haven't got it back since. Today I started again with the mainnet.2024-02-29.tar.gz database snapshot, but there's still no progress.

I generally use the latest Github version, but I reverted to the 2.1.4 release as per the warnings on newer releases. (This is still from Github, the source packages of releases don't build as they complain about not being Git repos, but that's another issue.)

### Version

git 9d0748fb988838cc783189094743aca6e4f3d440

### What platform are you using?

Linux

### What ui mode are you using?

CLI

### Relevant log output

```shell
2024-03-02T00:24:16.398 full_node chia.full_node.full_node: INFO     Requesting weight proof from peer 115.133.105.244 up to height 5018956
2024-03-02T00:24:20.586 full_node full_node_server        : INFO     Connection closed: 223.167.169.92, node id: 60f3e1cd067c47a280a630c9d8bf0cb93c1a5dc358d443145888109e55b19829
2024-03-02T00:24:20.587 full_node chia.full_node.full_node: INFO     peer disconnected PeerInfo(_ip=IPv4Address('223.167.169.92'), _port=0)
2024-03-02T00:24:25.119 wallet wallet_server              : INFO     Cannot connect to host 2409:8a55:3430:dc10:a4f8:3cac:8d12:327:8444 ssl:<ssl.SSLContext object at 0x7fa17da4dbe0> [Network is unreachable]
2024-03-02T00:24:25.681 full_node chia.full_node.weight_proof: INFO     validating 13070 sub epochs
2024-03-02T00:24:27.047 full_node chia.full_node.weight_proof: INFO     validate weight proof peak height 5018956
2024-03-02T00:24:27.066 full_node chia.full_node.weight_proof: ERROR    failed weight proof sub epoch sample validation
2024-03-02T00:24:27.075 full_node full_node_server        : WARNING  Banning 115.133.105.244 for 600 seconds
2024-03-02T00:24:27.075 full_node full_node_server        : INFO     Connection closed: 115.133.105.244, node id: 1c40e31a4ef2896335c7915617968872249d13648912f75cac34910811e94ccc
2024-03-02T00:24:27.076 full_node chia.full_node.full_node: INFO     peer disconnected PeerInfo(_ip=IPv4Address('115.133.105.244'), _port=8444)
2024-03-02T00:24:27.078 full_node chia.full_node.full_node: ERROR    Error with syncing: <class 'ValueError'>Traceback (most recent call last):
  File "/home/teknohog/sources/chia-blockchain/chia/full_node/full_node.py", line 975, in _sync
    fork_point, summaries = await self.request_validate_wp(
                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/teknohog/sources/chia-blockchain/chia/full_node/full_node.py", line 1040, in request_validate_wp
    raise ValueError("Weight proof validation failed")
ValueError: Weight proof validation failed
```

**Comment by teknohog** [comment]
Also seeing these signed/unsigned errors which could be related:

2024-03-02T15:25:20.566 full_node chia.full_node.full_node: ERROR    Error with syncing: <class 'ValueError'>Traceback (most recent call last):
  File "/home/teknohog/sources/chia-blockchain/chia/full_node/full_node.py", line 1034, in request_validate_wp
    validated, fork_point, summaries = await self.weight_proof_handler.validate_weight_proof(response.wp)
                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/teknohog/sources/chia-blockchain/chia/full_node/weight_proof.py", line 630, in validate_weight_proof
    valid, _ = await task
               ^^^^^^^^^^
  File "/home/teknohog/sources/chia-blockchain/chia/full_node/weight_proof.py", line 1652, in validate_weight_proof_inner
    if not validate_sub_epoch_sampling(rng, sub_epoch_weight_list, weight_proof):
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/teknohog/sources/chia-blockchain/chia/full_node/weight_proof.py", line 1583, in validate_sub_epoch_sampling
    weight_to_check = _get_weights_for_sampling(rng, tip.weight, weight_proof.recent_chain_data)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/teknohog/sources/chia-blockchain/chia/full_node/weight_proof.py", line 671, in _get_weights_for_sampling
    weight_to_check.append(uint128(int(weight)))
                           ^^^^^^^^^^^^^^^^^^^^
  File "/home/teknohog/sources/chia-blockchain/chia/util/struct_stream.py", line 71, in __init__
    raise ValueError(f"Value {self} does not fit into {type(self).__name__}")
ValueError: Value -20609 does not fit into uint128

**Comment by teknohog** [comment]
By the way, I've had to build my own chiapos library earlier to fix a buffer error – the binary libraries don't seem to like my Gentoo system. I now suspect some other library such as chiavdf is the issue here, but its build process looks more involved (it doesn't find my GMP library/headers, for starters).

I wish there were a nice unified way to build this from sources, instead of the current setup that downloads a bunch of binaries.

**Comment by emlowe** [comment]
To build entirely from source, including all the wheels, you can do `pip install -e .[dev] --no-binary :all:`

However, this is going to require a large number of dependencies including rust and is only viable for people who are comfortable with building binaries with many different systems.

**Comment by baerrs** [comment]
I'm getting this error as well, any updates?

**Comment by teknohog** [comment]
I've now found and fixed the error, it was due to bad CFLAGS in my Gentoo system. I removed -mfpmath=sse,387 from the flags and rebuilt Glibc, and now the node is syncing.

It's worth noting that things were fine until some late revisions of Glibc 2.38. I caught the issue as I noticed weird floating point results in Python, for example

>>> math.log(math.exp(1))
1.0018138846048894

which should equal 1.0.
